### PR TITLE
serverless refactor

### PR DESCRIPTION
**Pull request by DannyDouglass**
the driving factor for this pr was to refactor the deploy from `sam-cli` to `serverless`. along the way I added a few other items:
* automate `cognito openid provider` in `serverless`
* automate `cognito openid provider attribute mapping` in `serverless`
* added `provisionedConcurrency` to keep lambdas warm

**before merging**: 
- [x] add secrets to github action

## Related issues
* Fixes https://linear.app/issue/C-451/speed-up-github-sso-lambdas
* Fixes https://linear.app/issue/C-486/github-sso-deploy-broken

## Checklists
- [x] Make sure to link ticket
- [x] Make sure announce/discuss any changes to the #compliance channel in slack

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -4,6 +4,16 @@
   "lockfileVersion": 1,
   "requires": true,
   "dependencies": {
+    "2-thenable": {
```
Comment (scarney81) [inline package-lock.json]: Should this be a `yarn.lock` file instead of a `package-lock.json` file?

File: package.json [hunk 2]
```diff
@@ -9,18 +9,20 @@
   },
   "scripts": {
     "build": "webpack",
-    "test": "jest --runInBand --coverage",
-    "test-dev": "jest --runInBand --watch",
-    "start": "webpack --watch --display errors-only",
+    "build-dist": "npm run build",
```
Comment (scarney81) [inline package.json]: Maybe just call this `build`?

File: serverless.yml [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+  callback:
+    handler: src/connectors/lambda/callback.handler
+    events:
+      - http:
+          path: /callback
+          method: get
+          cors: true
+    environment:
+      DYNAMODB_STATE_TABLE: !Ref CognitoStatesTable
+      REGION: !Ref "AWS::Region"
```
Comment (scarney81) [inline serverless.yml]: You can get rid of these refs by adding a serverless plugin (if you want): https://serverless.com/plugins/serverless-pseudo-parameters/

File: .gitignore [hunk 4]
```diff
@@ -10,3 +10,11 @@ config.sh
...
 serverless-output.yml
 package-lock.json
+
+# Serverless directories
+.serverless
+.webpack
+
+npm-debug*
+
+config-dev.yml
```
Comment (scarney81) [inline .gitignore]: Might want to consider just supporting `.env` files. Then you either pump variables in or not - while avoiding all of the console warnings with serverless. Just a suggestion.

**Review by scarney81 — COMMENTED** [review]
Left some comments. Feel free to ping me if you want to discuss.

File: serverless.yml (hunk 3 above)
Comment (DannyDouglass) [reply serverless.yml]: good catch - this was a remanent. updated to use `${self:provider.region}` without the plugin.

File: package.json (hunk 2 above)
Comment (DannyDouglass) [reply package.json]: `build` script is above this puppy

File: package-lock.json (hunk 1 above)
Comment (DannyDouglass) [reply package-lock.json]: the open source path has been with npm. i'll make that adjustment as we build a new alternative here to open source.

File: .gitignore (hunk 4 above)
Comment (DannyDouglass) [reply .gitignore]: good suggestion, going to make this adjustment.

File: .env.development.example [hunk 5]
```diff
@@ -0,0 +1,4 @@
+AWS_USER_POOL_ID=""         # ex: us-east-1_2JEld13fj
+COGNITO_REDIRECT_URI=""     # ex: https://blah-123457890.auth.us-east-1.amazoncognito.com/oauth2/idpresponse"
+GITHUB_CLIENT_ID=""         # ex: 12kj4292j2n2k29d
+GITHUB_CLIENT_SECRET=""     # ex: badde13k3k39fk292nnwwgd1g1g220017
```
Comment (DannyDouglass) [inline .env.development.example:4]: note: these are not real values

File: .github/workflows/deploy.yml [hunk 6]
```diff
@@ -22,31 +22,14 @@ jobs:
...
-            sam deploy \
-              --capabilities CAPABILITY_IAM \
-              --debug TRUE \
-              --parameter-overrides CognitoRedirectUriParameter="$COGNITO_REDIRECT_URI" GitHubClientIdParameter="$GITHUB_CLIENT_ID" GitHubClientSecretParameter="$GITHUB_CLIENT_SECRET" StageNameParameter="$STAGE" \
-              --region "$REGION" \
-              --role-arn "$STACK_ROLE_ARN" \
-              --stack-name "$STACK_NAME" \
-              --tags githubSsoRoleArn="$STACK_ROLE_ARN" \
-              --template-file "$OUTPUT_TEMPLATE_FILE"
+        run: npm run serverless:deploy
```
Comment (DannyDouglass) [inline .github/workflows/deploy.yml]: ~20 lines => 1 💥

File: .github/workflows/deploy.yml [hunk 7]
```diff
@@ -22,31 +22,14 @@ jobs:
...
+          AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          AWS_USER_POOL_ID: ${{ secrets.AWS_USER_POOL_ID }}
           COGNITO_REDIRECT_URI: ${{ secrets.COGNITO_REDIRECT_URI }}
           GITHUB_CLIENT_ID: ${{ secrets.GITHUB_CLIENT_ID }}
           GITHUB_CLIENT_SECRET: ${{ secrets.GITHUB_CLIENT_SECRET }}
-          OUTPUT_TEMPLATE_FILE: serverless-output.yml
           NODE_ENV: production
           REGION: us-east-1
-          STACK_NAME: ${{ secrets.STACK_NAME }}
-          STACK_ROLE_ARN: ${{ secrets.STACK_ROLE_ARN}}
```
Comment (DannyDouglass) [inline .github/workflows/deploy.yml:51]: some of these were only needed for the old `sam-cli` deploy approach

File: package.json [hunk 8]
```diff
@@ -32,6 +34,7 @@
     "express": "^4.16.3",
     "json-web-key": "^0.3.0",
     "jsonwebtoken": "^8.3.0",
+    "serverless-dotenv-plugin": "^2.4.2",
```
Comment (DannyDouglass) [inline package.json:37]: @scarney81 using this serverless plugin to handle dotenv local configuration

File: src/connectors/web/routes.js [hunk 9]
```diff
@@ -8,5 +8,6 @@ module.exports = app => {
   app.get('/authorize', handlers.authorize);
   app.post('/authorize', handlers.authorize);
   app.get('/jwks.json', handlers.jwks);
+  app.get('/.well-known/jwks.json', handlers.jwks);
```
Comment (DannyDouglass) [inline src/connectors/web/routes.js:11]: bug fix from upstream repo

File: src/connectors/lambda/util/auth.js [hunk 10]
```diff
@@ -36,9 +36,10 @@ module.exports = {
       }
     }),
 
-  getIssuer: (host, stage) => {
-    const lStage = stage;
-    const issuer = `${host}/${lStage}`;
+
+  getIssuer: (host, requestContext) => {
+    const issuerPath = requestContext && requestContext.path.replace(requestContext.resourcePath, '') || '';
+    const issuer = `${host}${issuerPath}`;
```
Comment (DannyDouglass) [inline src/connectors/lambda/util/auth.js:42]: bug fix from upstream repo

File: .gitignore (hunk 4 above)
Comment (DannyDouglass) [reply .gitignore]: done via: https://serverless.com/plugins/serverless-dotenv-plugin/

File: serverless.yml (hunk 3 above)
Comment (DannyDouglass) [reply serverless.yml]: updated

**Review by scarney81 — APPROVED** [review]

**Review by troygoode — APPROVED** [review]

File: package-lock.json (hunk 1 above)
Comment (DannyDouglass) [reply package-lock.json]: fyi - decided to transition to yarn in the end per suggestion above

**Review by tone81 — APPROVED** [review]

**Pull request merged**